module/zfs/zfs_dir.c

```diff
@@ -746,9 +749,10 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
 			mutex_exit(&zp->z_lock);
 			return (SET_ERROR(ENOENT));
 		}
-		zp->z_links++;
+		inc_nlink(ZTOI(zp));
```

Comment:
The security hole is not an issue because we already check for z_unlinked, or you can use ZNEW to decide whether you want inc_nlink for set_nlink.